Refactor and update PURL matching in the API

another-rex opened the pull request:
Main functional change in this PR is making `arch=source` to mean `arch=<anything>`, since a source vulnerability will affect all architectures. 

Fixes a problem discussed here: #1321 

Did a bit of refactoring to do this:

- Moved the PURL matching logic around so that PURLs are not cleaned when passed around, but only cleaned when matching specifically with the datastore. 

- Also added the same additional matching step of `_match_purl` to `query_package` that's being used in `query_version`. 

- Updated `_match_purl` to consider `arch=source` to mean `arch=<anything>`.

oliverchang reviewed (approved):
nice! are there any tests we can add ?

another-rex commented:
Updated the integration test to test for PURLs with arch=source when queries have arch=amd64, also added a unit test to server.py for the `_match_purl()` function

oliverchang reviewed (approved):
thanks for adding the tests!!!
